src/coreclr/vm/util.hpp

```diff
@@ -71,6 +79,64 @@ BOOL inline FitsInU4(unsigned __int64 val)
     return val == (unsigned __int64)(unsigned __int32)val;
 }
 
+#if defined(DACCESS_COMPILE)
+#define FastInterlockedCompareExchange InterlockedCompareExchange
+#define FastInterlockedCompareExchangeAcquire InterlockedCompareExchangeAcquire
+#define FastInterlockedCompareExchangeRelease InterlockedCompareExchangeRelease
+#else
+
+#if defined(TARGET_WINDOWS) && defined(TARGET_ARM64)
+
+FORCEINLINE LONG  FastInterlockedCompareExchange(
+    LONG volatile *Destination,
+    LONG Exchange,
+    LONG Comperand)
+{
+    if (g_arm64_atomics_present)
+    {
+        return (LONG) __casal32((unsigned __int32*) Destination, (unsigned  __int32)Comperand, (unsigned __int32)Exchange);
```

Comment:
`__casal32` is an intrinsic exposed by VC++ only. For osx-arm64, we can just patch `PAL_ArmInterlockedOperationBarrier` to not call `__sync_synchronize`, right?

Reply:
> For osx-arm64, we can just patch PAL_ArmInterlockedOperationBarrier to not call __sync_synchronize, right?

Ah, no, that will depend on compilers flag, e.g. if someone adds `-mcpu=generic` to CMake then clang will stop emitting e.g. `casal` for us on osx-arm64 and the barrier will be needed. While an intrinsic would always be fine since there is no pre 8.1 hw on osx-arm64 (and will never be judging by apple's clang)

Reply:
> if someone adds `-mcpu=generic` to CMake

What is the default flag used for osx? Since its basline is 8.1, it should be `-mcpu=armv8.1` by default, even if we don't specify it? 

> While an intrinsic would always be fine since there is no pre 8.1 hw on osx-arm64

I don't understand how/why will someone want to use  `-mcpu=generic` and even if they use it, it will be `v8.1` only. Correct me if I am wrong.

> While an intrinsic would always be fine 

Can you check what code we generate today with default flags on osx for `__sync_val_compare_and_swap`? Should be `casal`. In that case, we can eliminate the `__sync_synchronize` for OSX. What change you tried that you mentioned in https://github.com/dotnet/runtime/issues/67824#issuecomment-1094337248?

Reply:
> What change you tried that you mentioned in https://github.com/dotnet/runtime/issues/67824#issuecomment-1094337248?

I think it was `-mcpu=native` (I verified coreclr via objdump) without that explicit barrier for all Interlocked functions including the one is GC and measured impact on TE:
![image](https://user-images.githubusercontent.com/523221/174631146-ee3eedf1-ccd7-4fa2-b0b7-44119c1d3535.png)

> What is the default flag used for osx? 

We don't set march/mcpu/m* flags so we use defaults

Reply:
- Use `-mcpu=native`. I don't think there is any disadvantage with this and in fact, Arm documentation [prefers to have that](https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/compiler-flags-across-architectures-march-mtune-and-mcpu).  Given that it was [added in 2018](https://github.com/llvm-mirror/clang/commit/86c991513001535af6b82bcb1f7c45ab60d2adf0), it should be available in clang9, but its availability is questionable for osx, see [this](https://stackoverflow.com/questions/65966969/why-does-march-native-not-work-on-apple-m1) and [this](https://github.com/UZ-SLAMLab/ORB_SLAM3/issues/542).
- For osx, we should definitely remove the barriers from `pal.h` because the baseline for M1 is `v8.1+` AFAIK. @VSadov - any objections?

Reply:
> For osx, we should definitely remove the barriers from pal.h because the baseline for M1 is v8.1+ AFAIK.

The barrier is a workaround for `__sync_val_compare_and_swap` not being a full fence. 

M1 is v8.1+, so `casal` would be sufficient and additional fence is redundant. 
but does compiler actually use `casal`?

Reply:
> but does compiler actually use casal

It does. https://godbolt.org/z/8xb3x7o71

I will put a PR out.

Reply:
if `-mcpu=generic` somehow flows in it will stop emitting casal

Reply:
https://github.com/dotnet/runtime/pull/71026

Reply:
Does it use `casal` in debug? Can it switch to old helper because of register pressure or if the old implementation is found faster (it could be).  
It feels like inline asm could have more reliable guarantees.